### ISSUE TITLE
IDE0001: False Positive for default primary constructor

### ISSUE THREAD

**Issue by AliveDevil** [dotnet/roslyn#75026]
<details>
<summary>

#### Version used: `dotnet --info`

</summary>

```
.NET SDK:
 Version:           8.0.400
 Commit:            36fe6dda56
 Workload version:  8.0.400-manifests.56cd0383
 MSBuild version:   17.11.3+0c8610977

Laufzeitumgebung:
 OS Name:     Windows
 OS Version:  10.0.22631
 OS Platform: Windows
 RID:         win-x64
 Base Path:   C:\Program Files\dotnet\sdk\8.0.400\

Installierte .NET-Workloads:
Konfiguriert für die Verwendung loose manifests beim Installieren neuer Manifeste.
Es sind keine installierten Workloads zum Anzeigen vorhanden.

Host:
  Version:      8.0.8
  Architecture: x64
  Commit:       08338fcaa5

.NET SDKs installed:
  3.1.426 [C:\Program Files\dotnet\sdk]
  5.0.408 [C:\Program Files\dotnet\sdk]
  6.0.202 [C:\Program Files\dotnet\sdk]
  6.0.321 [C:\Program Files\dotnet\sdk]
  8.0.400 [C:\Program Files\dotnet\sdk]

.NET runtimes installed:
  Microsoft.AspNetCore.App 3.1.25 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 3.1.32 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 5.0.17 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 6.0.4 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 6.0.5 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 6.0.12 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 6.0.26 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 6.0.33 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 8.0.8 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 3.1.25 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 3.1.32 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 5.0.17 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 6.0.4 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 6.0.5 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 6.0.12 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 6.0.15 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 6.0.26 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 6.0.33 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 8.0.8 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.WindowsDesktop.App 3.1.25 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 3.1.32 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 5.0.17 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 6.0.4 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 6.0.5 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 6.0.12 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 6.0.15 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 6.0.26 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 6.0.33 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 8.0.8 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]

Other architectures found:
  x86   [C:\Program Files (x86)\dotnet]
    registered at [HKLM\SOFTWARE\dotnet\Setup\InstalledVersions\x86\InstallLocation]

Environment variables:
  Not set

global.json file:
  Not found

Learn more:
  https://aka.ms/dotnet/info

Download .NET:
  https://aka.ms/dotnet/download
```
</details>

**Steps to Reproduce**:
Class, with default constructor:
```
// A
public abstract class Base<TOther>;
public sealed class Concrete() : Base<Concrete.Impl>() // IDE0001: Simplify Name
{
	public class Impl;
}

// B
public abstract class Base<TOther>();
public sealed class Concrete() : Base<Concrete.Impl>() // IDE0001: Simplify Name
{
	public class Impl;
}
```

Record class, with argument:
```
public abstract record class Base<TOther>(object A);
public sealed record class Concrete() : Base<Concrete.Impl>(default!) // IDE0001: Simplify Name
{
	public class Impl;
}
```

**Diagnostic Id**: `IDE0001: Simplify Name`

**Expected Behavior**: No false-positive, just like this does not produce a false-positive:
```
public abstract class Base<TOther>();

public sealed class Concrete() : Base<Concrete.Impl>
{
	public class Impl;
}
```

**Actual Behavior**: IDE0001 false-positive is emitted for `: Base<Concrete.Impl>()`, where it shouldn't.

**Comment by Orachor** [comment]
I'll try fixing this :)

**Comment by CyrusNajmabadi** [comment]
Awesome. Thanks!

**Comment by Orachor** [comment]
This one turned out more complicated than I expected. The issue is with the following code:

```
public abstract class C<T>;

public sealed class D() : C<E>()
{
    public class E;
}
```

Where the generic parameter `E` inside the base list is getting incorrectly bound even though members of D are not accessible in this context.

From what I have seen this case should be taken into account (according to the [comments here](https://github.com/dotnet/roslyn/blob/df4ae6b81013ac45367372176b9c3135a35a7e3c/src/Compilers/CSharp/Portable/Binder/BinderFactory.BinderFactoryVisitor.cs#L767)) however, in the end, the binder that is used is one that contains an `InContainerBinder` (the container being type D) which obviously binds the symbol.

I am still not 100% sure about some of the details as well as not knowing which part of the code needs to be changed (guessing it's the binder lookup), so if anyone wants to take over or has any ideas to help then that would be appreciated :)

**Comment by CyrusNajmabadi** [comment]
I don't have a suggestion for you.  It sounds like a binding issue in teh compiler.  You might want to talk to @dotnet/roslyn-compiler to see if htey ahve any ideas.  If you make a small repro test that demonstrates the problem when calling into the compiler apis, they may be amenable to fixing it.

**Comment by 333fred** [comment]
I'm not sure what you mean by "Where the generic parameter E inside the base list is getting incorrectly bound even though members of D are not accessible in this context." This isn't happening from what I can tell. As an example, this test passes:

```cs
        [Fact]
        public void QuickTest()
        {
            var source = """
                public abstract class C<T>;

                public sealed class D() : C<E>()
                {
                    public class E;
                }
                """;

            var comp = CreateCompilation(source);
            comp.VerifyDiagnostics(
                // (3,29): error CS0246: The type or namespace name 'E' could not be found (are you missing a using directive or an assembly reference?)
                // public sealed class D() : C<E>()
                Diagnostic(ErrorCode.ERR_SingleTypeNameNotFound, "E").WithArguments("E").WithLocation(3, 29)
            );

            var tree = comp.SyntaxTrees[0];
            var model = comp.GetSemanticModel(tree);
            var classD = tree.GetRoot().DescendantNodes().OfType<ClassDeclarationSyntax>().Skip(1).First();
            var cBaseList = classD.BaseList.Types[0].Type;
            var eArg = ((GenericNameSyntax)cBaseList).TypeArgumentList.Arguments[0];
            var e = model.GetSymbolInfo(eArg).Symbol;
            Assert.Null(e);
        }
```

Could you please clarify what you're seeing?

**Comment by AlekseyTs** [comment]
Is this problem really specific to primary constructors?

**Comment by AliveDevil** [comment]
Yes:
<details>

```
namespace Roslyn75026;

public abstract class Base<TOther>;

public sealed class PrimaryToBasePrimary() : Base<PrimaryToBasePrimary.Impl>()
{
    public class Impl;
}

public sealed class PrimaryToBaseDefault() : Base<PrimaryToBaseDefault.Impl>
{
    public class Impl;
}

public sealed class DefaultToBaseDefault : Base<DefaultToBaseDefault.Impl>
{
    public DefaultToBaseDefault()
    {
    }

    public class Impl;
}

public sealed class DefaultToBase : Base<DefaultToBase.Impl>
{
    public DefaultToBase() : base()
    {
    }

    public class Impl;
}

public sealed class ImplicitToBaseDefault : Base<ImplicitToBaseDefault.Impl>
{
    public class Impl;
}
```

</details>

Only the PrimaryToBasePrimary creats the IDE0001 suggestion.

**Comment by AliveDevil** [comment]
This btw doesn't affect default-primaries only, but all primaries, i.e.:
```
public abstract class Base<TOther>(string A);
public sealed class Concrete() : Base<Concrete.Impl>("A")
{
  public class Impl;
}
```

also has the IDE0001 annotation, but can only be worked around by discarding the primary constructor altogether.

**Comment by AlekseyTs** [comment]
It is not obvious why IDE is getting confused. If it gets an unexpected result from Compiler's API, we (Compiler team) need more details:
- what API
- under what conditions
- etc.

**Comment by Orachor** [comment]
The issue can be reproduced with this test:

```C#
  [Fact]
  public void Test()
  {
      var source = """
              public abstract class C<T>;

              public sealed class D() : C<D.E>() // Starting with working code
              {
                  public class E;
              }
              """;

      var comp = CSharpTestBase.CreateCompilation(source);

      var tree = comp.SyntaxTrees[0];
      var model = comp.GetSemanticModel(tree);
      var classD = tree.GetRoot().DescendantNodes().OfType<ClassDeclarationSyntax>().Skip(1).First();
      var cBaseList = classD.BaseList!.Types[0].Type;
      var deArg = ((GenericNameSyntax)cBaseList).TypeArgumentList.Arguments[0];

      var ceArg = SyntaxFactory.GenericName(SyntaxFactory.Identifier("C"), SyntaxFactory.TypeArgumentList([SyntaxFactory.IdentifierName("E")]));
      var simplified = SyntaxFactory.PrimaryConstructorBaseType(ceArg);

      model.TryGetSpeculativeSemanticModel(deArg.SpanStart, simplified, out var speculativeModel); // Replace C<D.E>() with C<E>()

      var eArg = ((GenericNameSyntax)simplified.Type).TypeArgumentList.Arguments[0];
      var e = speculativeModel!.GetSymbolInfo(eArg).Symbol;
      Assert.Null(e); // throws, e is D.E here
  }
```

I am not 100% sure but I'm assuming that the speculative semantic model is on the IDE side of things, so perhaps there are no issues with `GetSymbolInfo` and it's just a matter of incorrectly using the API (since the original example works correctly).

**Comment by AlekseyTs** [comment]
The used overload of ```TryGetSpeculativeSemanticModel``` is documented as "Get a SemanticModel object that is associated with a constructor initializer ..."
What this meant to say - it is about the executable part of the node, i.e. the argument list. The type part is not even part of the binding process, any changes to the base type portion of the node are completely ignored. No matter what that part say, the initializer part is bound in context of the same constructor declared by the original syntax. For the purpose of that speculative semantic model, the base type doesn't change and a possibility of it changing isn't even considered as something to pay attention to.
When the symbol info is requested, we determine that the node is not bound as part of the initializer. It is bound in an error recovery mode, as a stand-alone node in context of the initializer, where the nested type is in scope. I guess, we could tighten behavior of the speculative semantic model so that it doesn't return any semantic information for nodes outside of the argument list. However, that probably won't help the goal that IDE is trying to achieve in this scenario. I think IDE should use a "regular" way of speculating on base types within base list. Something like:
```
model.GetSpeculativeSymbolInfo(deArg.SpanStart, SyntaxFactory.IdentifierName("E"), SpeculativeBindingOption.BindAsTypeOrNamespace)
```    
Or whatever mechanism is used when argument list is not present.

**Comment by Orachor** [comment]
Thanks for the in depth explanation, now I understand 🙂
